src/ir/auth_logic/ast.h

```diff
@@ -139,6 +139,55 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { kNumberType, kPrincipalType };
+
+// Relation declaration types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArgumentType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArgumentType(Kind kind, std::string name)
```

Comment:
No need for `std::move` for kind. Kind is already cheap to copy.

optional: Instead of taking a string, you could take an absl::string_view:
```C++
explicit ArgumentType(Kind kind, absl::string_view name): kind_(kind), name_(name) {}
...
absl::string_view name() const { return name_; }

```
Note that  the field `  std::string name_; ` would still stay as string.